File: CardMasterSkin/Skins/SkinElement.vb
﻿Imports CardMasterCard.Card
Imports CardMasterSkin.GraphicsElement
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.IO

Namespace Skins

    Public MustInherit Class SkinElement

        Private Enum TextureTypes
            Color
            GradientColor
            Image
        End Enum

        Public Property X As Integer = 0
        Public Property Y As Integer = 0
        Public Property Width As Integer = 0
        Public Property Height As Integer = 0
        Public Property Shadow As SkinShadow = Nothing

        Private m_color1 As Color = Nothing
        Private m_color2 As Color = Nothing
        Private m_imageName As String = Nothing
        Private m_texturesDirectory As DirectoryInfo = Nothing
        Private m_type As TextureTypes

        Protected m_graphics As Graphics = Nothing

        Protected MustOverride Function GetGraphicElements(card As Card) As List(Of GraphicElement)

        Protected Sub New(width As Integer, height As Integer)
            Me.New(0, 0, width, height)
        End Sub

        Protected Sub New(x As Integer, y As Integer, width As Integer, height As Integer)
            Me.X = x
            Me.Y = y
            Me.Width = width
            Me.Height = height

            SetBackground(Color.Black)

        End Sub

        Public Sub Draw(g As Graphics, card As Card)

            m_graphics = g

            Dim graphicElements As List(Of GraphicElement) = GetGraphicElements(card)

            For Each graphicElement As GraphicElement In graphicElements
                graphicElement.DrawShadow(g, Shadow)
                graphicElement.Draw(g)
            Next

            m_graphics = Nothing

        End Sub

        Public Sub SetBackground(color As Color)
            m_color1 = color
            m_color2 = Nothing
            m_texturesDirectory = Nothing
            m_imageName = Nothing
            m_type = TextureTypes.Color
        End Sub

        Public Sub SetBackground(color1 As Color, color2 As Color)
            m_color1 = color1
            m_color2 = color2
            m_texturesDirectory = Nothing
            m_imageName = Nothing
            m_type = TextureTypes.GradientColor
        End Sub

        Public Sub SetBackground(texturesDirectory As DirectoryInfo, imageName As String)
            m_color1 = Nothing
            m_color2 = Nothing
            m_texturesDirectory = texturesDirectory
            m_imageName = imageName
            m_type = TextureTypes.Image
        End Sub

        Public Function GetBackground() As Brush

            Dim bkg As Brush = Nothing

            Select Case m_type
                Case TextureTypes.Color
                    bkg = New SolidBrush(m_color1)

                Case TextureTypes.GradientColor
                    bkg = New LinearGradientBrush(New Point(X, 1), New Point(Height, 1), m_color1, m_color2)

                Case TextureTypes.Image
                    Dim files() As FileInfo = m_texturesDirectory.GetFiles(m_imageName & ".*", SearchOption.AllDirectories)

                    If files.Count > 0 Then
                        Dim img As Image = New Bitmap(files.First.FullName)
                        bkg = New TextureBrush(img, WrapMode.Clamp)
                    End If

            End Select

            If bkg Is Nothing Then
                bkg = New SolidBrush(Color.Black)
            End If

            Return bkg

        End Function

    End Class

End Namespace

File: CardMasterSkin/Skins/SkinElement.cs
using CardMasterCard.Card;
using CardMasterSkin.GraphicsElements;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.IO;

namespace CardMasterSkin.Skins
{
    public abstract class SkinElement
    {
        private static Dictionary<String, Brush> imgTexturesMap = new Dictionary<String, Brush>();

        private enum TextureTypes : int { Color, GradientColor, Image };

        public int X { get; set; } = 0;
        public int Y { get; set; } = 0;
        public int Width { get; set; } = 0;
        public int Height { get; set; } = 0;
        public SkinElementShadow Shadow { get; set; } = null;
        public SkinElementBorder Border { get; set; } = null;
        public bool Visible { get; set; } = true;
        protected Color color1;
        protected Color color2;
        protected string imageName = null;
        private TextureTypes type;
        protected Skin skin = null;


        protected Graphics graphics = null;

        protected abstract List<GraphicElement> GetGraphicElements(Card card);

        protected SkinElement(Skin skin, int width, int height) : this(skin, 0, 0, width, height)
        { }

        protected SkinElement(Skin skin, int x, int y, int width, int height)
        {
            this.X = x;
            this.Y = y;
            this.Width = width;
            this.Height = height;
            this.skin = skin;

            SetBackground(Color.Black);
        }

        ~SkinElement()
        {
            this.Shadow = null;
            this.skin = null;
            this.graphics = null;
        }

        public void Draw(Graphics g, Card card)
        {
            this.graphics = g;

            List<GraphicElement> graphicElements = GetGraphicElements(card);

            foreach(GraphicElement graphicElement in graphicElements)
            {
              
                    if (this.Shadow != null)
                    {
                        graphicElement.DrawShadow(g, this.Shadow);
                    }

                    graphicElement.Draw(g);

                    if (this.Border != null)
                    {
                        graphicElement.DrawBorder(g, this.Border);
                    }
               
            }

            graphicElements.Clear();
            graphicElements = null;

            this.graphics = null;
        }

        public void SetBackground(Color color)
        {
            this.color1 = color;
            this.imageName = null;
            this.type = TextureTypes.Color;
        }

        public void SetBackground(Color color1, Color color2)
        {
            this.color1 = color1;
            this.color2 = color2;
            this.imageName = null;
            this.type = TextureTypes.GradientColor;
        }

        public void SetBackground(String imageName)
        {
            this.imageName = imageName;
            this.type = TextureTypes.Image;
        }

        public Brush GetBackground()
        {
            Brush bkg = null;

            switch (this.type)
            {
                case TextureTypes.Color:
                    bkg = new SolidBrush(this.color1);
                    break;

                case TextureTypes.GradientColor:
                    bkg = new LinearGradientBrush(new Point(X, 1), new Point(Height, 1), this.color1, this.color2);
                    break;

                case TextureTypes.Image:
                    if ((this.skin.TexturesDirectory != null) && (this.skin.TexturesDirectory.Exists))
                    {
                        FileInfo[] files = this.skin.TexturesDirectory.GetFiles(this.imageName + ".*", SearchOption.AllDirectories);

                        if (files.Length > 0)
                        {
                            string fileName = files[0].FullName;

                            try
                            {
                                bkg = imgTexturesMap[fileName];
                            }
                            catch (KeyNotFoundException)
                            {
                                bkg = null;
                            }

                            if (bkg == null)
                            {
                                Image img = new Bitmap(fileName);
                                //bkg = new TextureBrush(img, WrapMode.Clamp);
                                bkg = new TextureBrush(img, WrapMode.TileFlipXY);
                                img.Dispose();
                                img = null;

                                imgTexturesMap[fileName] = bkg;

                            }
                        }

                        files = null;
                    }
                    break;
            }

            if (bkg == null)
            {
                bkg = new SolidBrush(Color.Black);
            }

            return bkg;
        }

    }
}
